matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt

```diff
@@ -1,920 +1,918 @@
+            val sessionId = restoredOutboundGroupSession.outboundGroupSession.sessionIdentifier()
+            // cache it
+            outboundGroupSessionCache[sessionId] = GroupSessionCacheItem(roomId, restoredOutboundGroupSession.outboundGroupSession)
+
+            return MXOutboundSessionInfo(
+                    sessionId = sessionId,
+                    sharedWithHelper = SharedWithHelper(roomId, sessionId, store),
+                    restoredOutboundGroupSession.creationTime
+            )
+        }
+        return null
+    }
+
+    fun discardOutboundGroupSessionForRoom(roomId: String) {
+        val toDiscard = outboundGroupSessionCache.filter {
+            it.value.groupId == roomId
+        }
+        toDiscard.forEach { (sessionId, cacheItem) ->
+            cacheItem.groupSession.releaseSession()
+            outboundGroupSessionCache.remove(sessionId)
+        }
+        store.storeCurrentOutboundGroupSessionForRoom(roomId, null)
+    }
+
+    /**
+     * Get the current session key of  an outbound group session.
+     *
+     * @param sessionId the id of the outbound group session.
+     * @return the base64-encoded secret key.
+     */
+    fun getSessionKey(sessionId: String): String? {
+        if (sessionId.isNotEmpty()) {
+            try {
+                return outboundGroupSessionCache[sessionId]!!.groupSession.sessionKey()
+            } catch (e: Exception) {
+                Timber.tag(loggerTag.value).e(e, "## getSessionKey() : failed")
+            }
+        }
+        return null
+    }
+
+    /**
+     * Get the current message index of an outbound group session.
+     *
+     * @param sessionId the id of the outbound group session.
+     * @return the current chain index.
+     */
+    fun getMessageIndex(sessionId: String): Int {
+        return if (sessionId.isNotEmpty()) {
+            outboundGroupSessionCache[sessionId]!!.groupSession.messageIndex()
+        } else 0
+    }
+
+    /**
+     * Encrypt an outgoing message with an outbound group session.
+     *
+     * @param sessionId     the id of the outbound group session.
+     * @param payloadString the payload to be encrypted and sent.
+     * @return ciphertext
+     */
+    fun encryptGroupMessage(sessionId: String, payloadString: String): String? {
+        if (sessionId.isNotEmpty() && payloadString.isNotEmpty()) {
+            try {
+                return outboundGroupSessionCache[sessionId]!!.groupSession.encryptMessage(payloadString)
+            } catch (e: Throwable) {
+                Timber.tag(loggerTag.value).e(e, "## encryptGroupMessage() : failed")
+            }
+        }
+        return null
+    }
+
+    //  Inbound group session
+
+    sealed class AddSessionResult {
+        data class Imported(val ratchetIndex: Int) : AddSessionResult()
+        abstract class Failure : AddSessionResult()
+        object NotImported : Failure()
+        data class NotImportedHigherIndex(val newIndex: Int) : AddSessionResult()

```

Comment:
should it inherit `Failure` instead of `AddSessionResult` ?
Also you could use `sealed interface` instead of `sealed class`.

Reply:
done